### ISSUE TITLE
Investigate how to map XPath expressions onto AST nodes

### ISSUE THREAD

**Issue by MEZk** [checkstyle/checkstyle#4369]
We need  to investigate how to map XPath expressions onto AST nodes. To begin with, it is worth designing a model of mapper which will allow to map XPath expressions onto AST nodes. The main idea of the mapper is to connect XPath expressions which will be based on the values of [TokenTypes](https://github.com/MEZk/checkstyle/blob/dabe377fd60374da3c5cfaddc8126847468b65ef/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java#L34-L34) enum onto the nodes of DetailAST. 

For example, if we have the following code fragment:
```java
public class TestClass {
    void method() {
    }
}
```

the following tree structure:
```xml
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|  `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
`--OBJBLOCK -> OBJBLOCK [1:23]
    |--LCURLY -> { [1:23]
    |--METHOD_DEF -> METHOD_DEF [2:4]
    |  |--MODIFIERS -> MODIFIERS [2:4]
    |  |--TYPE -> TYPE [2:4]
    |  |  `--LITERAL_VOID -> void [2:4]
    |  |--IDENT -> method [2:9]
    |  |--LPAREN -> ( [2:15]
    |  |--PARAMETERS -> PARAMETERS [2:16]
    |  |--RPAREN -> ) [2:16]
    |  `--SLIST -> { [2:18]
    |      `--RCURLY -> } [3:4]
    `--RCURLY -> } [4:0]
```

and the following XPath expression (example, format should be discussed):
```xml
//ClassDef[@Text='TestClass']
```

than the parser should connect the XPath expression elements with the AST nodes in this way:

```
DetailAST (type: CLASS_DEF, text: "TestClass") <- XpathExpression(ClassDef, text: "TestClass")
```

The base idea can be taken from PMD. They implemented both Saxon and Jaxen mappers. From my point of view, Saxon mapper is easier to reimplement for our needs. [Saxon XPath rule mapper](https://github.com/pmd/pmd/blob/d92688a07dc5e6edc3bfc81cced6f25c951a19f4/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/SaxonXPathRuleQuery.java#L40-L40) maps each XPath node onto AST node. Their algorithm is based on [the tree traversal algorithm](https://github.com/pmd/pmd/blob/d92688a07dc5e6edc3bfc81cced6f25c951a19f4/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/xpath/saxon/ElementNode.java#L38-L38). They [map](https://github.com/pmd/pmd/blob/d92688a07dc5e6edc3bfc81cced6f25c951a19f4/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/xpath/saxon/ElementNode.java#L46-L46) each node of the AST starting with root onto XPath element.

**Thus, what should be done in the scope of this issue:**
1. Design XPath expression format for the new Checkstyle Suppression Model;
2. Investigate what library (Saxon, Jaxen, or native Java) should be used to implement mapper;
3. Provide base details of mapper model implementation (UML diagram of classes can be a good example).

**Comment by timurt** [comment]
@MEZk 
I am agree with you, I have tried both Jaxen and Saxon and noticed some moments. 
1. Both implementations can work on DOM, XOM, DOM4j and JDOM, but Saxon also supports JDOM2 and AXIOM tree models
2. Jaxen project is probably not active now, last changes on Git repository were about 2 years ago (https://github.com/jaxen-xpath/jaxenand official site is unavailable (http://jaxen.codehaus.org/, www.jaxen.org/), while Saxon's last release was in 5 April 2017.
3. Jaxen does not support XPATH 2.0, while Saxon does
4. PMD uses Jaxen for own suppression model. They also allow to select Jaxen/Saxon for writing custom rules.
I think Saxon library is more reliable and has more functionalities comparing to Jaxen.

**Comment by vanniktech** [comment]
How about not mapping it to AST and instead map it to UAST (Universal Abstract Syntax Tree)? That way also adding support for Kotlin, which is an official thing for Android now, can be supported.

**Comment by timurt** [comment]
@MEZk 

> Design XPath expression format for the new Checkstyle Suppression Model;

I compared AST components between PMD and Checkstyle . I think using constants from [TokenTypes](https://github.com/MEZk/checkstyle/blob/dabe377fd60374da3c5cfaddc8126847468b65ef/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java#L34-L34) as parts for XPath queries is not good idea. For example `CTOR_DEF` is not so obvious for Constructor declaration and `LITERAL_while` contains underscore,lowercase and uppercase characters which is difficult for user. I like PMD format, all AST components names consist of description and category: `ASTAnnotationMethodDeclaration`, `ASTPreIncrementExpression`, `ASTBreakStatement` and etc. 

Having full names in camel-case format instead of shorten tokens will make Xpath queries more understandable I guess. Here some mapping examples:
CTOR_DEF -> ConstructorDeclaration
METHOD_DEF -> MethodDeclaration
VARIABLE_DEF -> VariableDeclaration
LITERAL_if -> IfStatement
LITERAL_while -> WhileStatement

Here some queries examples:
`//CLASS_DEF[@Text='TestClass']/METHOD_DEF/SLIST/LITERAL_if`
vs
`//ClassDeclaration[@Text='TestClass']/MethodDeclaration/StatementsList/IfStatement`

**Comment by MEZk** [comment]
@timurt 
1) LITERAL_while is not a constant from TokenTypes enum. 
2) ClassDeclaration, MethodDeclaration, StatementsList, are not attributes of DetailAST, but TokenTypes are. In order to connect them to DetailAST we will need to implement converter to bind them to one of the attributes of DetailAST.
3) PMD uses the names as their tree nodes have are called in that way. They have an hierarchy of AST nodes, we don't. The nodes in our AST can be identified by token types.

**Comment by MEZk** [comment]
@vanniktech 
> How about not mapping it to AST and instead map it to UAST (Universal Abstract Syntax Tree)? That way also adding support for Kotlin, which is an official thing for Android now, can be supported.

Checkstyle works only with AST as we depend on ANTLR and java grammar.

**Comment by MEZk** [comment]
The main idea of the mapper is to connect XPath expressions which will be based on the values of TokenTypes enum onto the nodes of DetailAST.  We can use Saxon library to work with XPath expressions. XPath expressions consists of document nodes. Each node can be mapped to corresponding AST node. Thus, we need to implement our own XPath-AST document-based tree model. Saxon offers an interface NodeInfo which represents a node in Saxon's implementation of the XPath 2.0 data model. This is the primary interface for accessing trees in Saxon, and it forms part of the public Saxon API. In our case, we need to design our own implementation which will combine the information about XPath and AST nodes. We also need the implementation of the root (document) node which will contain the reference to the root of XPath-AST document-based tree. 
The class XPathExpression from Saxon represents XPath expression which can be executed, returning the result as a List, whose members will be instances of the class which implements NodeInfo interface. 
To perform AST tree traversal with XPath expression we need to extend the Navigator class from Saxon in order to apply its algorithms to XPath-AST document-based tree. The Navigator class provides helper classes for navigating a tree, irrespective of its implementation.
 Each AST node has its attributes (methods which access fields information). XPath expression can use the attributes to refer to AST node attributes. To iterate over the attributes the Navigator subclass is required.

Documentation:
http://www.saxonica.com/html/documentation9.6/expressions/range.html
http://saxon.sourceforge.net/saxon7.6/api-guide.html

**Comment by romani** [comment]
> and the following XPath expression (example, format should be discussed): //ClassDef[@Text='TestClass']

It should be `//CLASS_DEF[@text='TestClass']` , TokenType is used as name ,all attributes are properties from DetailAST.

> Jaxen project is probably not active now, last changes on Git repository were about 2 years ago

yes, lets avoid it.

> LITERAL_while contains underscore,lowercase and uppercase characters which is difficult for use

it is ` public static final int LITERAL_WHILE = ...` - [code](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java#L1665)

>  I like PMD format, all AST components names consist of description and category: ASTAnnotationMethodDeclaration, ASTPreIncrementExpression, ASTBreakStatement and etc.

It is there AST structure and name - so they do not invent new names too, what is used in development that is used in xpath.
it is verbose, and verbose is good when smb read this text on daily basis, or need to create expression without help of any tool. Creation of suppression in PMD in very hard process. I still a problem for to make new suppression for PMD and I always use existing examples of their tree tool to collect name of nodes.
They do have problems with names too - `@Image`
PMDs suppressions are long - https://github.com/checkstyle/checkstyle/blob/master/config/pmd.xml#L33 , additional pain is too have several path is one string - very hard to maintain. It will be very good if user can make several suppress lines in his config to keep long xpath separately and easy to read.

Our user already aware of AST tokens, they configure Checks by them, they already know about AST , they do not know exact structure (nobody in the world can predict AST tree), so we need a tool that will generate xpath for user to copy paste in configuration.

If DetailAST extend, we do not need to change our xpath processor/matcher, as it will just for with it as with java bean.

**Comment by MEZk** [comment]
@romani 
> It should be //CLASS_DEF[@text='TestClass'] , TokenType is used as name ,all attributes are properties from DetailAST.

Yes, your are right. It was my first attempt to show the format. The Xpath will be constructed based on TokenType enum values.

> it is public static final int LITERAL_WHILE = ... - code

Yes. Should be LITERAL_WHILE.

> They do have problems with names too - @Image

It was their design mistake, as for me, to save node name in the field with name "image".

> additional pain is too have several path is one string - very hard to maintain. It will be very good if user can make several suppress lines in his config to keep long xpath separately and easy to read

We can allow to have as many ```suppression-xpath``` elements as user wants.

> Our user already aware of AST tokens, they configure Checks by them, they already know about AST 

That was my main point. 

> Our user already aware of AST tokens, they configure Checks by them, they already know about AST 

Yes. We will have problems only if the AST structure is changed. For example, if METHOD_DEF will become the parent of CLASS_DEF :), etc.

**Comment by timurt** [comment]
@MEZk @rnveach 
Sorry for delay, was struggling with Saxon Navigator/NodeInfo implementation. 
Here is [My commit](https://github.com/timurt/checkstyle/commit/86aa0153779cd893ebf79eef616a0fb1c6f8dd2c) where I have created simple Saxon xpath implementation with tests. 

The realization is pretty simple: 
`Node` - general interface
`AbstractNode` - abstract class implements `Node`
`DocumentNode` - actually this is not node, it is info about document
`RootNode` - class represents root node, extends `AbstractNode`
`ElementNode` - class represents `ClassDefinition` node

Is this correct way?

**Comment by MEZk** [comment]
@timurt 
How are you going to connect the nodes with AST nodes? Just TokenType is not enough.

**Comment by timurt** [comment]
@MEZk 
Added DetailNode and DetailAst fields to node
Implemented `[@Image='SomeClass']`
Added javadoc

**Comment by MEZk** [comment]
@timurt 
First implementation can support only direct xpath-expressions. For example, there might be two ways of suppressing a violation for the method 'foo'.

```java
public class TestClass {
    void foo() {
    }
}
```

```
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|  `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
`--OBJBLOCK -> OBJBLOCK [1:23]
    |--LCURLY -> { [1:23]
    |--METHOD_DEF -> METHOD_DEF [2:4]
    |  |--MODIFIERS -> MODIFIERS [2:4]
    |  |--TYPE -> TYPE [2:4]
    |  |  `--LITERAL_VOID -> void [2:4]
    |  |--IDENT -> foo [2:9]
    |  |--LPAREN -> ( [2:15]
    |  |--PARAMETERS -> PARAMETERS [2:16]
    |  |--RPAREN -> ) [2:16]
    |  `--SLIST -> { [2:18]
    |      `--RCURLY -> } [3:4]
    `--RCURLY -> } [4:0]
```

1) /CLASS_DEF//OBJBLOCK//METHOD_DEF[@text='foo']
2) METHOD_DEF[@text='foo']

First implementation of suppression model can handle only case 1. We will add support of case 2 later.

**Comment by timurt** [comment]
@MEZk 
I maybe wrong but I am not agree with the second one

> 1. /CLASS_DEF//OBJBLOCK//METHOD_DEF[@text='foo']  


- we can query it from any node because it has full path from the root

> 2. METHOD_DEF[@text='foo']

- context-based query, so we should query from `OBJBLOCK` node, if we query from the root it should return empty array, no?

3. //METHOD_DEF[@text='foo']  - this query will get all possible method definitions

**Comment by timurt** [comment]
@MEZk 
Here is my corrected commit [4369](https://github.com/timurt/checkstyle/commit/fd98bc916a1b4d9eb3b748777a701cd8409c96dd)

If I am not wrong only thing to implement left, is attribute iterator, to make queries like this
`//CLASS_DEF[@*]` or `*[@text]`

**Comment by MEZk** [comment]
@timurt 
Good step forward. I will review the changes ASAP.

**Comment by MEZk** [comment]
@timurt 
Reviewed the changes.

**Comment by timurt** [comment]
@MEZk 
New commit  [4369](https://github.com/timurt/checkstyle/commit/73128750635bf64289b02d5451448efb4fa86e83)
resolved bugs and moments you mentioned

Additionally 
1. added attribute iterator
2. resolved problem with querying symbolic characters like `TokenTypes.RCURLY, TokenTypes.LCURLY, TokenTypes.SLIST` and etc.
3. Added UTs to test `Attribute query, Parent query, And query` and etc

**Comment by MEZk** [comment]
@timurt 
Good, need some time to review.

**Comment by MEZk** [comment]
@timurt 
I reviewed changes. You did a good job.
1) Please, fix/answer the points mentioned.
2) Please, prepare the commit for the PR, fix checkstyle/PMD/... violations. Use ```mvn clean verify``` on your local.
3) Check that your changes satisfy coverage threshold. Use ```mvn clean cobertura:check```.
4) When you are ready, please notify me here, I will do the final review and you can send PR.

**Comment by timurt** [comment]
@MEZk 
Here is my [commit](https://github.com/timurt/checkstyle/commit/c67e983cb93e0677350a8a8647632c14370591e7)

It passed mvn verification and cobertura checks.

I have question regarding `DetailAst`, how can I get string representation of token type, for example
`TokenTypes.BLOCK_COMMENT_BEGIN -> "BLOCK_COMMENT_BEGIN"`, 
`DetailAst.getText()` sometimes returns actual value, for example
`STRING_LITERAL -> "HelloWorld"` or `ARRAY_DECLARATOR -> [`
We can write switch statements for every token type, but is there more simple another way?

**Comment by MEZk** [comment]
@timurt 
Unfortunately, TokenType is not an enum, it is just a class with lots of constants. We cannot use valueOf, etc methods. We have two options here: 
1) Implement a logic which will convert string representation (getText) to the TokenType. Similar to switch-case.
2) Use reflection to get the name of the field.

@romani 
What we can do in this situation? For [some token types](https://github.com/timurt/checkstyle/commit/c67e983cb93e0677350a8a8647632c14370591e7#diff-baa604e0025ec3fb63e7a5fd3851eed0R84) ```getText``` method returns the string representation of the token (RCURLY -> }, SLIST -> }, etc). We use ```getText``` method to refer to the node attribute ```text```, for example CLASS_DEF[@text='MyClass'].

**Comment by romani** [comment]
> Use reflection to get the name of the field.

Lets try this approach if we could do this ones only, to create a  `Map<int,String>`  to map integer value to string.

**Comment by MEZk** [comment]
@timurt 
Please, try to use reflection to get the name of the field by token type.

**Comment by MEZk** [comment]
@timurt 
Ping

**Comment by timurt** [comment]
@MEZk 
Here is my [commit](https://github.com/timurt/checkstyle/commit/5f9909bcf70a37d125e7dc5995e7b769eda9dbd8) with reflection

I am trying to suppress PMD violations for `AbstractNode`, but when I use `@SuppressWarnings` annotation I get this message:

> AbstractNode.java:52:19: The warning 'PMD' cannot be suppressed at this location. If you try to suppress IntelliJ IDEA inspection, please use javadoc block tag @noinspection [SuppressWarnings]

What can be the problem?

**Comment by romani** [comment]
> What can be the problem?

`@SuppressWarnings` is not allowed in our code as it is pollute code  and all nuances/problems of incpection/checks/rules/ ...... should be configured outside of sources. IDEA inspections do not have such feature but they do have suppression in javadoc, that is preferable.
PMD suppressions are stored in PMD config files by xpath expressions.

**Comment by MEZk** [comment]
@timurt 
> Here is my commit with reflection

I did final review.

1) Please, fix new points.
2) Please, prepare your changes for PR. They should pass at least ```mvn clean verify``` on your local.
3) Send PR.

**Comment by rnveach** [comment]
Fix was merged

**Comment by romani** [comment]
I marked issue as "misc" as feature for full functional Xpath Filter is not ready to be viewed by users.
